### PR TITLE
Multi validator service

### PR DESCRIPTION
**Pull request by mkalinin**
Resolves https://github.com/ethereum/ethereumj/issues/1189

### CODE REVIEW

File: ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java [hunk 1]
```diff
@@ -68,17 +75,25 @@ public ProposerServiceImpl(BeaconProposer proposer, BeaconChain beaconChain,
...
+
+        for (byte[] pubKey : pubKeys) {
+            Validator validator = state.getValidatorSet().getByPubKey(pubKey);
+            if (validator != null) {
+                this.validatorIndices.add(validator.getIndex());
+                this.pubKeysMap.put(validator.getIndex(), validator.getPubKey());
+            } else {
+                // something went wrong
+                logger.error("Failed to start proposer for {}: validator does not exist", HashUtil.shortHash(pubKey));
+                return;
```
Comment (zilm13) [inline ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: I'd either throw an exception on this or continue to parse everything the rest. Both would be better, I guess.

File: ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java [hunk 2]
```diff
@@ -68,17 +75,25 @@ public ProposerServiceImpl(BeaconProposer proposer, BeaconChain beaconChain,
...
+            }
 
-        Validator validator = state.getValidatorSet().getByPupKey(config.pubKey());
-        // something went wrong
-        if (validator == null) {
-            logger.error("Failed to start proposer: validator {} does not exist", Hex.toHexString(config.pubKey()));
-            return;
+            this.validatorIndices.add(validator.getIndex());
```
Comment (zilm13) [inline ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: why this one is needed?

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/db/TrieValidatorSet.java [hunk 3]
```diff
@@ -95,7 +94,7 @@ public synchronized Validator get(Integer index) {
 
     @Nullable
     @Override
-    public Validator getByPupKey(byte[] pubKey) {
```
[the remark this answers is not in the document]
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/db/TrieValidatorSet.java]: Addressed here https://github.com/ethereum/ethereumj/pull/1205/files#diff-e608ccea27ce12512de530106ad08952R52

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java [hunk 4]
```diff
@@ -73,22 +75,27 @@ public int getShardId() {
         /* length of validators array */
         private final int committeeSize;
         /* index in validators array */
-        private final int validatorIdx;
+        private final int arrayIdx;
```
[the remark this answers is not in the document]
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java]: It's an index in `Committee.validators[]` array. I am going crazy with all this indices too 🤪 Would love to see better structure for committees.

File: ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java (hunk 1 above)
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: Agree on that! It's too silent

File: ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java (hunk 2 above)
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: I mean
```
(86th line)                 this.validatorIndices.add(validator.getIndex());
(94th line)                 this.validatorIndices.add(validator.getIndex());
```
Looks strange in diff

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java]: so what's the structure finally?
[validatorSet (commitee) per slot][validator in validatorSet]

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/db/TrieValidatorSet.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/db/TrieValidatorSet.java]: Just funny how did we miss it.

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java]: The structure is:
```
  slot[]    // each slot cross-links on a number of shards
   \
    committee[]    // each committee has a validators array
     \
      validators[]    // numbers in this array correspond to validator set indices, 
                      // indices in this array are defined by shuffling, 
                      // and this is what arrayIdx does hold
```

File: ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java (hunk 2 above)
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: Good catch 👍

Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: @zilm13 What do you think about this update https://github.com/ethereum/ethereumj/commit/a2030c4261e13d1eae49ec8f5758def0cb9fc65b?

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/db/TrieValidatorSet.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/db/TrieValidatorSet.java]: Ah, yes :) It reads pretty similar. `b` is just flipped `p` 😆

File: ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java (hunk 2 above)
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/sharding/proposer/ProposerServiceImpl.java]: looks like redundancy removed, cool!

File: ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java]: Could you move this cool ascii image to the Javadoc?

[the remark this answers is not in the document]
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/sharding/processing/state/Committee.java]: Done https://github.com/ethereum/ethereumj/pull/1195/files#diff-9d2b5214c8604b3dcff4347fd7d0ac0cR67

**Pull request merged**